subprojects/logging/src/main/java/org/gradle/internal/logging/LoggingCommandLineConverter.java

```diff
@@ -18,36 +18,28 @@
 
 import com.google.common.collect.BiMap;
 import com.google.common.collect.HashBiMap;
-import org.apache.commons.lang.StringUtils;
 import org.gradle.api.logging.LogLevel;
-import org.gradle.api.logging.configuration.ConsoleOutput;
 import org.gradle.api.logging.configuration.LoggingConfiguration;
 import org.gradle.api.logging.configuration.ShowStacktrace;
 import org.gradle.cli.AbstractCommandLineConverter;
 import org.gradle.cli.CommandLineArgumentException;
 import org.gradle.cli.CommandLineParser;
 import org.gradle.cli.ParsedCommandLine;
+import org.gradle.internal.buildoption.BuildOption;
 
 import java.util.Arrays;
 import java.util.HashSet;
-import java.util.Locale;
-import java.util.Map;
+import java.util.List;
 import java.util.Set;
 
 public class LoggingCommandLineConverter extends AbstractCommandLineConverter<LoggingConfiguration> {
-    public static final String DEBUG = "d";
-    public static final String DEBUG_LONG = "debug";
-    public static final String WARN = "w";
-    public static final String WARN_LONG = "warn";
-    public static final String INFO = "i";
-    public static final String INFO_LONG = "info";
-    public static final String QUIET = "q";
-    public static final String QUIET_LONG = "quiet";
-    public static final String CONSOLE = "console";
-    public static final String FULL_STACKTRACE = "S";
-    public static final String FULL_STACKTRACE_LONG = "full-stacktrace";
-    public static final String STACKTRACE = "s";
-    public static final String STACKTRACE_LONG = "stacktrace";
+    private final List<BuildOption<LoggingConfiguration>> buildOptions = new LoggingConfigurationBuildOptionFactory().create();
+    public static final String DEBUG = LoggingConfigurationBuildOptionFactory.DebugOption.SHORT_OPTION;
+    public static final String WARN = LoggingConfigurationBuildOptionFactory.WarnOption.SHORT_OPTION;
+    public static final String INFO = LoggingConfigurationBuildOptionFactory.InfoOption.SHORT_OPTION;
+    public static final String QUIET = LoggingConfigurationBuildOptionFactory.QuietOption.SHORT_OPTION;
+    public static final String FULL_STACKTRACE = LoggingConfigurationBuildOptionFactory.FullStacktraceOption.SHORT_OPTION;
```

Comment:
same here, these 2 are different command-line options that affect a single build option - the stack trace style.

Reply:
I like the idea of providing more stack trace display options in the future.

Reply:
@adammurdoch How were you envisioning the API for the stack traces as one build option? I see the following two differences that doesn't make it fit the convenience types we have right now.

1. You'll need to be able to pass in multiple instances of `CommandLineOptionConfiguration`.
2. You'll need to have access to the `ParsedCommandLine` to determine the `ShowStacktrace`.

We might have to implement a subtype of `AbstractBuildOption` for this.

```
public static class StacktraceOption extends AbstractBuildOption<LoggingConfiguration> {
    public static final String STACKTRACE_LONG_OPTION = "stacktrace";
    public static final String STACKTRACE_SHORT_OPTION = "s";
    public static final String FULL_STACKTRACE_LONG_OPTION = "full-stacktrace";
    public static final String FULL_STACKTRACE_SHORT_OPTION = "S";

    public StacktraceOption() {
        super(null, CommandLineOptionConfiguration.create(STACKTRACE_LONG_OPTION, STACKTRACE_SHORT_OPTION, "Print out the stacktrace for all exceptions."), CommandLineOptionConfiguration.create(FULL_STACKTRACE_LONG_OPTION, FULL_STACKTRACE_SHORT_OPTION, "Print out the full (very verbose) stacktrace for all exceptions."));
    }

    @Override
    public void applyFromProperty(Map<String, String> properties, LoggingConfiguration settings) {
        // not supported
    }

    @Override
    public void configure(CommandLineParser parser) {
        for (CommandLineOptionConfiguration config : commandLineOptionConfigurations) {
            CommandLineOption option = parser.option(config.getAllOptions())
                .hasDescription(config.getDescription())
                .deprecated(config.getDeprecationWarning());

            if (config.isIncubating()) {
                option.incubating();
            }
        }
    }

    @Override
    public void applyFromCommandLine(ParsedCommandLine options, LoggingConfiguration settings) {
        if (options.hasOption(STACKTRACE_LONG_OPTION)) {
            settings.setShowStacktrace(ShowStacktrace.ALWAYS);
        } else if (options.hasOption(FULL_STACKTRACE_LONG_OPTION)) {
            settings.setShowStacktrace(ShowStacktrace.ALWAYS_FULL);
        }
    }
}
```